### DISCUSSION TITLE
Transaction API

### DISCUSSION THREAD

**Discussion by johnbcodes** [vrischmann/zig-sqlite#137] [Ideas]
I'd like to add transaction functionality and ergonomics similar to the savepoint API. Thoughts?

Here's an API sketch:

```zig
pub fn transaction(self: *Self) Transaction.InitError!Transaction {
    return self.transactionWithBehavior(TransactionBehavior.Deferred);
}

pub fn transactionWithBehavior(self: *Self, behavior: TransactionBehavior) Transaction.InitError!Transaction {
    return Transaction.init(self, behavior);
}

pub const TransactionBehavior = enum {
    Deferred,
    Immediate,
    Exclusive,
};

// Implementation very similar to Savepoint
pub const Transaction = struct {
    fn init(db: *Db, behavior: TransactionBehavior) Error!Self {
        //...
    }
    pub fn commit(self: *Self) void {
        //...
    }
    // No statement execution if committed
    pub fn rollback(self: *Self) void {
        //...
    }
};
```

**Comment by vrischmann** [comment]
Hi,

I'm fine adding a transaction API, your sketch looks good to me.

I have never added transactions because I was thinking savepoints were strictly superior since they allow nesting, but it looks like they don't allow specifying IMMEDIATE or EXCLUSIVE.

**Comment by johnbcodes** [comment]
I happen to have a use for `IMMEDIATE` which is why I asked :smile:

Now though I'm concerned about `commit()` and `rollback()` not returning errors if their statements fail. My experience tells me that it would lead to situations where code/processes would continue normally when otherwise they would fail fast if they knew of the error. There are at least two downsides to changing their signature if changes to `Savepoint` were included. First it's a breaking API change and second the idiom of `defer tx.rollback()` no longer works. It seems that `defer try tx.rollback()` isn't allowed because returns aren't allowed in `defer` statements/blocks. As much as I like the ergonomics of `defer tx.rollback()` I'd prefer the flexibility of knowing about the error and my own choices on how to handle it.

**Comment by johnbcodes** [comment]
After some time and tinkering I feel less sure about `rollback()` returning errors, at least in non-nested transactions.  I struggle to come up with scenarios where I would do something more with a rollback error than add it to telemetry/log it or ignore it. A callback could be added in the future to preserve `defer tx.rollback()` and allow for more than the current logging.

I don't have experience with nested transactions so I don't know if there are scenarios where a failed rollback has more impact.

Still feel strongly about `commit()` returning an error though. It would stil be a breaking change for `Savepoint` but all around it would be a smaller one.

**Comment by johnbcodes** [comment]
Ugh.. thinking through error conditions hurts my brain. To be fair, it can get complicated. So, if the rollback statement fails with with certain errors the [documentation](https://www.sqlite.org/lang_transaction.html) seems to suggest that the transaction can remain open. This could cause similar unwanted database state to commits that didn't actually happen. So... I guess I'm back to wanting to handle the errors.. I think.

Apologies for all the noise.

**Comment by vrischmann** [comment]
No worries about the noise. I'm sorry I don't have much input though, right now I don't really have much time or energy to put into this project.
